simuvex/plugins/symbolic_memory.py

```diff
@@ -649,132 +650,228 @@ def _store(self, req):
-                cv = sv
-            elif req.condition is not None and num_addresses == 1:
-                cv = self.state.se.If(req.condition, sv, fv)
-            elif req.condition is None and num_addresses != 1:
-                cv = self.state.se.If(req.addr == a, sv, fv)
-            elif req.condition is not None and num_addresses != 1:
-                cv = self.state.se.If(self.state.se.And(req.addr == a, req.condition), sv, fv)
+        l.debug("... done")
+        req.completed = True
+        return req
 
-            req.conditional_values.append(cv)
+    def _insert_memory_object(self, value, address, size):
+        value.make_uuid()
+        self.state.scratch.dirty_addrs.update(range(address, address+size))
+        mo = SimMemoryObject(value, address, length=size)
+        self.mem.store_memory_object(mo)
 
-        if type(req.addr) not in (int, long) and req.addr.symbolic:
-            conditional_constraint = self.state.se.Or(*[ req.addr == a for a in req.actual_addresses ])
-            if (conditional_constraint.symbolic or  # if the constraint is symbolic
-                    conditional_constraint.is_false()):  # if it makes the state go unsat
-                req.constraints.append(conditional_constraint)
+    def _store_fully_concrete(self, address, size, data, endness, condition):
+        assert address is not None and not self.state.solver.symbolic(address)
+        assert size is not None and not self.state.solver.symbolic(size)
 
-        #
-        # now simplify
-        #
+        size = self.state.solver.any_int(size)
+        if size < data.length/8:
+            data = data[size*8-1:]
+        address = self.state.solver.any_int(address)
+        original_value = self._read_from(address, size)
+        if endness == "Iend_LE" or (endness is None and self.endness == "Iend_LE"):
+            original_value = original_value.reversed
+        conditional_value = self.state.solver.If(condition, data, original_value)
 
-        if (self.category == 'mem' and options.SIMPLIFY_MEMORY_WRITES in self.state.options) or \
-           (self.category == 'reg' and options.SIMPLIFY_REGISTER_WRITES in self.state.options):
-            req.simplified_values = [ self.state.se.simplify(cv) for cv in req.conditional_values ]
-        else:
-            req.simplified_values = list(req.conditional_values)
+        return [ dict(value=conditional_value, addr=address, size=size) ]
 
-        #
-        # fix endness
-        #
 
-        if req.endness == "Iend_LE" or (req.endness is None and self.endness == "Iend_LE"):
-            req.stored_values = [ sv.reversed for sv in req.simplified_values ]
-        else:
-            req.stored_values = list(req.simplified_values)
+    def _store_symbolic_size(self, address, size, data, endness, condition):
+        assert address is not None and not self.state.solver.symbolic(address)
+        assert size is not None and self.state.solver.symbolic(size)
```

Comment:
`assert all(v[3] is write_list[0][3] for v in write_list)`  <- This assertion is kinda crucial, remove as well?